packages/runner/sync-customer-os-data/repository/interaction_event_repository.go

```diff
@@ -0,0 +1,224 @@
+	MergeInteractionSession(ctx context.Context, tenant string, date time.Time, message entity.EmailMessageData) (string, error)
+	MergeInteractionEvent(ctx context.Context, tenant string, date time.Time, message entity.EmailMessageData) (string, error)
+	MergeInteractionEventToSession(ctx context.Context, tenant string, interactionEventId, interactionSessionId string) error
+
+	InteractionEventSentByEmail(ctx context.Context, tenant, interactionEventId, emailId string) error
+	InteractionEventSentByUser(ctx context.Context, tenant, interactionEventId, userExternalId, externalSystem string) error
+
+	InteractionEventSentToEmails(ctx context.Context, tenant, interactionEventId, sentType string, emails []string) error
+}
+
+type interactionEventRepository struct {
+	driver *neo4j.DriverWithContext
+}
+
+func NewInteractionEventRepository(driver *neo4j.DriverWithContext) InteractionEventRepository {
+	return &interactionEventRepository{
+		driver: driver,
+	}
+}
+
+func (r *interactionEventRepository) MergeInteractionSession(ctx context.Context, tenant string, syncDate time.Time, message entity.EmailMessageData) (string, error) {
+	session := utils.NewNeo4jWriteSession(ctx, *r.driver)
+	defer session.Close(ctx)
+
+	query := "MERGE (is:InteractionSession_%s {identifier:$identifier, source:$source, channel:$channel}) " +
+		" ON CREATE SET " +
+		"  is:InteractionSession, " +
+		"  is.id=randomUUID(), " +
+		"  is.syncDate=$syncDate, " +
+		"  is.createdAt=$createdAt, " +
+		"  is.updatedAt=$updatedAt, " +
+		"  is.name=$name, " +
+		"  is.status=$status," +
+		"  is.type=$type," +
+		"  is.sourceOfTruth=$sourceOfTruth, " +
+		"  is.appSource=$appSource " +
+		" WITH is " +
+		" RETURN is.id"
+
+	dbRecord, err := session.ExecuteWrite(ctx, func(tx neo4j.ManagedTransaction) (any, error) {
+		queryResult, err := tx.Run(ctx, fmt.Sprintf(query, tenant),
+			map[string]interface{}{
+				"tenant":        tenant,
+				"source":        message.ExternalSystem,
+				"sourceOfTruth": message.ExternalSystem,
+				"appSource":     message.ExternalSystem,
+				"identifier":    message.EmailThreadId,
+				"name":          message.Subject,
+				"syncDate":      syncDate,
+				"createdAt":     message.CreatedAt,
+				"updatedAt":     message.CreatedAt,
+				"status":        "ACTIVE",
+				"type":          "THREAD",
+				"channel":       "EMAIL",
+			})
+		if err != nil {
+			return nil, err
+		}
+		record, err := queryResult.Single(ctx)
+		if err != nil {
+			return nil, err
+		}
+		return record, nil
+	})
+	if err != nil {
+		return "", err
+	}
+	return dbRecord.(*db.Record).Values[0].(string), nil
+}
+
+func (r *interactionEventRepository) MergeInteractionEvent(ctx context.Context, tenant string, syncDate time.Time, message entity.EmailMessageData) (string, error) {
+	session := utils.NewNeo4jWriteSession(ctx, *r.driver)
+	defer session.Close(ctx)
+
+	query := "MERGE (ie:InteractionEvent_%s {externalId:$externalId, source:$source, channel:$channel}) " +
```

Comment:
Would consider link with external system as in case of other entities since messages has unique external id